src/Magnum/Platform/EmscriptenApplication.cpp

```diff
@@ -0,0 +1,505 @@
+    //
+    // Map first touch as left mouse, yes or no?
+    /*
+    emscripten_set_touchstart_callback("#canvas", this, false,
+        [](int, const EmscriptenTouchEvent* event, void* userData) -> Int {
+            MouseEvent e{event};
+            reinterpret_cast<EmscriptenApplication*>(userData)->mousePressEvent(e);
+            return e.isAccepted();
+        });
+    emscripten_set_touchend_callback("#canvas", this, false,
+        [](int, const EmscriptenTouchEvent* event, void* userData) -> Int {
+            MouseEvent e{event};
+            reinterpret_cast<EmscriptenApplication*>(userData)->mouseReleaseEvent(e);
+            return e.isAccepted();
+        });
+    emscripten_set_touchmove_callback("#canvas", this, false,
+        [](int, const EmscriptenTouchEvent* event, void* userData) -> Int {
+            MouseMoveEvent e{event};
+            reinterpret_cast<EmscriptenApplication*>(userData)->mouseMoveEvent(e);
+            return e.isAccepted();
+        });
+    */
+
+    /* keypress_callback does not fire for most of the keys and the modifiers don't seem to work,
+      keydown on the other hand works fine for all */
+    auto keyDownCallback =
+        [](int, const EmscriptenKeyboardEvent* event, void* userData) -> Int {
+            EmscriptenApplication* app = reinterpret_cast<EmscriptenApplication*>(userData);
+            if(app->isTextInputActive() && strlen(event->key) == 1) {
+                TextInputEvent e{{event->key, 1}};
+                app->textInputEvent(e);
+                return e.isAccepted();
+            }
+            KeyEvent e{event};
+            app->keyPressEvent(e);
+            return e.isAccepted();
+        };
+    emscripten_set_keydown_callback("#canvas", this, false, keyDownCallback);
+
+    auto keyReleaseCallback =
+        [](int, const EmscriptenKeyboardEvent* event, void* userData) -> Int {
+            KeyEvent e{event};
+            reinterpret_cast<EmscriptenApplication*>(userData)->keyReleaseEvent(e);
+            return e.isAccepted();
+        };
+    emscripten_set_keyup_callback("#canvas", this, false, keyReleaseCallback);
+}
+
+EmscriptenApplication::KeyEvent::Key EmscriptenApplication::KeyEvent::toKey(const EM_UTF8* key, const EM_UTF8* code) {
+    const size_t keyLength = strlen(key);
+    if(keyLength == 0) return Key::Unknown;
+
+    /* We use key for a-z as it gives us a keyboard layout respecting
+       representation of the key, i.e. we get `z` for z depending on layout
+       where code may give us `y` independent of the layout. */
+    if(keyLength == 1) {
+        if(key[0] >= 'a' && key[0] <= 'z') return Key(key[0]);
+        else if(key[0] >= 'A' && key[0] <= 'Z') return Key(key[0] + 32);
+    }
+
+    /* We use code for 0-9 as it allows us to differentiate towards Numpad digits.
+       For digits independent of numpad or not, key is e.g. '0' for Zero */
+    if(Utility::String::beginsWith(code, "Digit")) {
```

Comment:
Unfortunately the `beginsWith()` etc. functions take a `std::string` as a first parameter, which means it gets allocated again on every call. I need to introduce a StringView class first to fix that, until then can you use `viewBeginsWith()` everywhere instead?